### ISSUE TITLE
Cannot configure dist_timeout when using device_mesh

### ISSUE THREAD

**Issue by mvpatel2000** [pytorch/pytorch#119574]
### 🐛 Describe the bug

When calling `dist.initialize_dist`, I can specify a `dist_timeout` argument. 

When training with FSDP and device_mesh, I want to call `from torch.distributed._tensor import init_device_mesh` and pass the `device_mesh` into FSDP. However, it seems that the process groups created do not respect `dist_timeout` through this process

https://github.com/pytorch/pytorch/blob/5d6e323549bd5d3997e8c344532e611078be7011/torch/distributed/device_mesh.py#L292

### Versions

Torch nightly 1/10/2024

**Comment by mvpatel2000** [comment]
Oops this is not distributed checkpointing 😢. Someone from torch end please feel free to remove label.

**Comment by wz337** [comment]
@wanchaol Want to surface this request to you. We can follow up to discuss how we can expose this option to users.

**Comment by wanchaol** [comment]
IMO it's a bit tricky as we want to support all possible pg options and still keep simple enough UX, it could be sth like this:

```
init_device_mesh(
    device_type: str,
    mesh_shape: Tuple[int],
    *,
    mesh_dim_names: Optional[Tuple[str, ...]] = None,
    mesh_dim_options: Optional[str, Any] = None
)
```
We need to do some study about current pg options I feel.

**Comment by wconstab** [comment]
>    mesh_dim_options: Optional[str, Any] = None

well, the simplest way to cover it would be for mesh_dim_options to accept 
`Optional[Union[List[Dict[str, Any]], Dict[str, Any]]]`
* user passes {timeout: blah}. -- applies to all created PGs
* user pages [{timeout: blah}, {timeout: baz}] - expect one dict per mesh dim 

dicts can have limited validation on what kwargs are supported, but can generally just be passed via **kwargs to the underlying init_process_group call.

on second thought maybe List is bad, and maybe Optional isn't necessary.  Instead what if we just do

`mesh_dim_options: Dict[Dict[str, Any]]` -- 

  {'dp': {timeout...}} -- only apply to dp mesh
  vs 
  {} -- empty, so use defaults

a little annoying to have to write the names of all the meshes if you want to set values that apply to all of them.  could add a "*" wildcard or some other way.

idk, do you think its better to manually support individual kwargs at the init_device_mesh layer?  One downside is that even if you only pick 2 or 3 args to support at init_device_mesh layer, you then have to expose the clunky per-dim-name vs global handling for each one of them, and on the user side it could get a bit repetitive.

**Comment by wz337** [comment]
> > mesh_dim_options: Optional[str, Any] = None
> 
> well, the simplest way to cover it would be for mesh_dim_options to accept `Optional[Union[List[Dict[str, Any]], Dict[str, Any]]]`
> 
> * user passes {timeout: blah}. -- applies to all created PGs
> * user pages [{timeout: blah}, {timeout: baz}] - expect one dict per mesh dim
> 
> dicts can have limited validation on what kwargs are supported, but can generally just be passed via **kwargs to the underlying init_process_group call.
> 
> on second thought maybe List is bad, and maybe Optional isn't necessary. Instead what if we just do
> 
> `mesh_dim_options: Dict[Dict[str, Any]]` --
> 
> {'dp': {timeout...}} -- only apply to dp mesh vs {} -- empty, so use defaults
> 
> a little annoying to have to write the names of all the meshes if you want to set values that apply to all of them. could add a "*" wildcard or some other way.
> 
> idk, do you think its better to manually support individual kwargs at the init_device_mesh layer? One downside is that even if you only pick 2 or 3 args to support at init_device_mesh layer, you then have to expose the clunky per-dim-name vs global handling for each one of them, and on the user side it could get a bit repetitive.

Ye. I agree that adding the option can make things clunky. 
Although we envisioned `init_device_mesh` to be the user-facing API instead of `init_process_group`, in this case, with more advanced use case, would it better to have users init the default process group to specify the options (such as timeout) before we figure out a decent solution to expose options in DeviceMesh side?

**Comment by wz337** [comment]
@wconstab We are also thinking of a new class method as `DeviceMesh.from_process_group()` that may be able to accommodate changing pg options as an alternative.

**Comment by wconstab** [comment]
I think to preserve layering and be flexible, users should get exactly the same options and abilities when they create their own PGs and wrap them into a mesh, or use the mesh to create PGs for them. 

This implies mesh init fn should accept same options as regular pg init would.

And if creating mesh from existing PGs, you'd have to support a list of existing PGs to build out the N-d mesh right? Or were you thinking of only passing the world PG and then mesh always creates its own PGs? That could work but I think in cases people already build their own 2 or 3D PGs, converting to use mesh would be easier if you let them use existing PGs

**Comment by mvpatel2000** [comment]
@wconstab @wz337 any chance this will be fixed soon?

**Comment by wanchaol** [comment]
@mvpatel2000 one question I have on this: curious if you need to control the dist timeout on every device mesh dimension, or simply the global timeout and every dimension use the same timeout?

If that's the later case, you can simply call `init_process_group` to init a world process group with the global timeout, before `init_device_mesh` (which init subgroups)

If it's the former case, we currently have some private API that allow this, haven't figure out the exact UX yet https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L162

**Comment by mvpatel2000** [comment]
The latter is fine -- I would like device mesh to respect the timeout I have set when initializing device. 

The bug report is to say I am attempting the latter, where I run `dist.init_process_group(device_obj.dist_backend, timeout=timeout_timedelta)` but device mesh does not respect the timeout I pass in and instead has a separate timeout.

**Comment by wanchaol** [comment]
> The latter is fine -- I would like device mesh to respect the timeout I have set when initializing device.
> 
> The bug report is to say I am attempting the latter, where I run `dist.init_process_group(device_obj.dist_backend, timeout=timeout_timedelta)` but device mesh does not respect the timeout I pass in and instead has a separate timeout.

Ohhh I see, this make a lot of sense! It could actually be a bug in our c10d layer, the new_group doesn't respect default group timeout, we should fix this

**Comment by wconstab** [comment]
hmm. i think it was working for me in torchtitan. can you share your exact repro @mvpatel2000?

https://github.com/pytorch/torchtitan/blob/040ea1db300ff968fc5e5d71f9b6b38af8e2e135/torchtitan/utils.py#L106

**Comment by mvpatel2000** [comment]
Will try to rerun on latest torch, this was originally reported on 1/10 so could have changed...

**Comment by Changlin-Lee** [comment]
Any updates to fix this problem?
